### PR TITLE
Change the OpenSSL initializer to enforce FIPS and enable SymCrypt test

### PR DESCRIPTION
**Pull request by xumia**
Change the OpenSSL initializer to enforce FIPS
Support test for SymCrypt Openssl
Add step to build OpenSSL with SymCrypt enabled in azure pipeline
Apply the patch similar for Mariner: https://github.com/microsoft/CBL-Mariner/blob/2.0/SPECS/openssl/openssl-1.1.1-fips-SymCrypt.patch
Disable some test cases not supported

### CODE REVIEW

File: src/openssl.patch/10-support-fips-mode.patch [hunk 1]
```diff
@@ -67,26 +137,30 @@ index 1b0d523bea..af171bda16 100644
...
+-            && !RUN_ONCE(&engine_dynamic, ossl_init_engine_dynamic))
+-        return 0;
++    if (opts & OPENSSL_INIT_ENGINE_DYNAMIC)
++    {
++        if (!RUN_ONCE(&engine_dynamic, ossl_init_engine_dynamic))
++            return 0;
++        RUN_ONCE(&fips_config, ossl_init_fips_conf);
++        if (g_fips_mode_enabled == 1)
++            RUN_ONCE(&engine_symcrypt, ossl_init_engine_symcrypt);
 +    }
```
Comment (samuel-lee-msft) [inline src/openssl.patch/10-support-fips-mode.patch:160]: I think you'll also want to add a constructor which guarantees this is called at libcrypto load time. See https://github.com/microsoft/CBL-Mariner/blob/0f7c7b22d05848e440458f3aa57a1866c78f7637/SPECS/openssl/openssl-1.1.1-load-default-engines.patch#L9-L14

Comment (samuel-lee-msft) [reply src/openssl.patch/10-support-fips-mode.patch:160]: Ah you put this in the 20-support-fips-test patch. As long as both patches are always applied it doesn't matter too much, but might be clearer for the default load behavior change to go here.

File: src/openssl.patch/debian.patch/20-support-fips-test.patch [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ 
+     setup_test_framework();
+ 
++    if (!SCOSSL_ENGINE_Initialize())
++    {
++        test_printf_stderr("SCOSSL init failed - aborting\n");
++        return ret;
++    }
++
++
```
Comment (samuel-lee-msft) [inline src/openssl.patch/debian.patch/20-support-fips-test.patch]: Given the patches to force SCOSSL load at libcrpyto load time, I don't think this or the scossl.h includes are necessary any more.

File: src/openssl.patch/debian.patch/40-disable-test-cases-with-fips-enabled.patch [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+-ok(run(test(["enginetest"])), "running enginetest");
+diff --git a/test/recipes/30-test_evp.t b/test/recipes/30-test_evp.t
+index 2385105b8b..c0b792d903 100644
+--- a/test/recipes/30-test_evp.t
++++ b/test/recipes/30-test_evp.t
+@@ -14,7 +14,7 @@ use OpenSSL::Test qw/:DEFAULT data_file/;
+ 
+ setup("test_evp");
+ 
+-my @files = ( "evpciph.txt", "evpdigest.txt", "evpencod.txt", "evpkdf.txt",
```
Comment (samuel-lee-msft) [inline src/openssl.patch/debian.patch/40-disable-test-cases-with-fips-enabled.patch]: Removing evpciph.txt is a step further than necessary. You can just remove tests for AES-GCM with IV length != 12 bytes (lines 780-804, 838-862, 896-920, and 960-984 last I checked)

**Review by samuel-lee-msft — CHANGES_REQUESTED** [review]

File: src/openssl.patch/debian.patch/40-disable-test-cases-with-fips-enabled.patch (hunk 3 above)
Comment (xumia) [reply src/openssl.patch/debian.patch/40-disable-test-cases-with-fips-enabled.patch]: Fixed, thanks.

File: src/openssl.patch/10-support-fips-mode.patch (hunk 1 above)
Comment (xumia) [reply src/openssl.patch/10-support-fips-mode.patch:160]: Moved the change to the file crypto/engine/eng_all.c from 20-support-fips-test patch to 10-support-fips-mode.patch.

File: src/openssl.patch/debian.patch/20-support-fips-test.patch (hunk 2 above)
Comment (xumia) [reply src/openssl.patch/debian.patch/20-support-fips-test.patch]: When I remove it, some tests failed with the error " /home/xumia/sonic-fips-2/src/openssl/build_static/engines/symcryptengine.so" not found. Can we keep the change?

**Review by samuel-lee-msft — APPROVED** [review]
I think this is looking good now, provided all tests are passing :)

**Pull request merged**